src/Workspaces/Remote/ServiceHub/Shared/ServiceHubServiceBase.cs

```diff
@@ -46,14 +33,76 @@ internal abstract class ServiceHubServiceBase : IDisposable
 
         private RoslynServices _lazyRoslynServices;
 
-        private bool _disposed;
-
         protected ServiceHubServiceBase(IServiceProvider serviceProvider, Stream stream)
             : this(serviceProvider, stream, SpecializedCollections.EmptyEnumerable<JsonConverter>())
         {
         }
 
         protected ServiceHubServiceBase(IServiceProvider serviceProvider, Stream stream, IEnumerable<JsonConverter> jsonConverters)
+            : base(serviceProvider, stream, jsonConverters)
+        {
+        }
+
+        protected RoslynServices RoslynServices
+        {
+            get
+            {
+                if (_lazyRoslynServices == null)
+                {
+                    _lazyRoslynServices = new RoslynServices(_solutionInfo.ScopeId, AssetStorage, RoslynServices.HostServices);
+                }
+
+                return _lazyRoslynServices;
+            }
+        }
+
+        protected Task<Solution> GetSolutionAsync(CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(_solutionInfo);
+
+            return GetSolutionAsync(RoslynServices, _solutionInfo, cancellationToken);
+        }
+
+        protected Task<Solution> GetSolutionAsync(PinnedSolutionInfo solutionInfo, CancellationToken cancellationToken)
+        {
+            var localRoslynService = new RoslynServices(solutionInfo.ScopeId, AssetStorage, RoslynServices.HostServices);
+            return GetSolutionAsync(localRoslynService, solutionInfo, cancellationToken);
+        }
+
+        public virtual void Initialize(PinnedSolutionInfo info)
+        {
+            // set pinned solution info
+            _lazyRoslynServices = null;
+            _solutionInfo = info;
+        }
+
+        private static Task<Solution> GetSolutionAsync(RoslynServices roslynService, PinnedSolutionInfo solutionInfo, CancellationToken cancellationToken)
+        {
+            var solutionController = (ISolutionController)roslynService.SolutionService;
+            return solutionController.GetSolutionAsync(solutionInfo.SolutionChecksum, solutionInfo.FromPrimaryBranch, solutionInfo.WorkspaceVersion, cancellationToken);
+        }
+    }
+
+    /// <summary>
+    /// Base type with helper methods. this is not tied to how Roslyn OOP works. just servicehub helpers.
+    /// </summary>
+    internal abstract class ServiceBase : IDisposable
+    {
+        private static int s_instanceId;
+
+        private readonly JsonRpc _rpc;
+
+        protected readonly int InstanceId;
+
+        protected readonly TraceSource Logger;
+        protected readonly AssetStorage AssetStorage;
+
+        [Obsolete("don't use RPC directly but use it through StartService and InvokeAsync", error: true)]
```

Comment:
Is this field used outside via IVT? If not, can we delete it instead of marking obsolete?

Reply:
Ah, you are just moving existing code to a new type. Seems reasonable to leave it as is in this PR.